File: modules/saga/src/main/scala/saga/app/SagaProducer.scala
package saga.app
import java.util.UUID

import model.messages.{ActionRequest, SagaResponse, SagaStateTransition}
import model.saga.Saga
import model.topics.{ActionTopic, SagaTopic}
import org.apache.kafka.streams.kstream.{KStream, Produced}

object SagaProducer {

  def actionRequests[A](ctx: SagaContext[A], actionRequests: KStream[UUID, ActionRequest[A]]): Unit = {
    actionRequests
      .to(ctx.aSpec.topicNamer(ActionTopic.request),
          Produced.`with`(ctx.aSpec.serdes.uuid, ctx.aSerdes.request))
  }

  def sagaState[A](ctx: SagaContext[A], sagaState: KStream[UUID, Saga[A]]): Unit = {
    sagaState.to(ctx.sSpec.topicNamer(SagaTopic.state),
                 Produced.`with`(ctx.sSpec.serdes.uuid, ctx.sSerdes.state))
  }

  def sagaStateTransitions[A](ctx: SagaContext[A],
                              stateTransitions: KStream[UUID, SagaStateTransition[A]]*): Unit = {
    stateTransitions.foreach(
      _.to(ctx.sSpec.topicNamer(SagaTopic.stateTransition),
           Produced.`with`(ctx.sSerdes.uuid, ctx.sSerdes.transition)))
  }

  def sagaResponses[A](ctx: SagaContext[A], sagaResponse: KStream[UUID, SagaResponse]) = {
    sagaResponse.to(ctx.sSpec.topicNamer(SagaTopic.response),
                    Produced.`with`(ctx.sSpec.serdes.uuid, ctx.sSerdes.response))
  }

}

File: modules/saga/src/main/java/io/simplesource/saga/saga/app/SagaProducer.java
package io.simplesource.saga.saga.app;

import java.util.UUID;

import io.simplesource.saga.model.messages.ActionRequest;
import io.simplesource.saga.model.messages.SagaResponse;
import io.simplesource.saga.model.messages.SagaStateTransition;
import io.simplesource.saga.model.saga.Saga;
import io.simplesource.saga.shared.topics.TopicTypes;
import org.apache.kafka.streams.kstream.KStream;
import org.apache.kafka.streams.kstream.Produced;


public final class SagaProducer {

    public static <A> void actionRequests(SagaContext<A> ctx, KStream<UUID, ActionRequest<A>> actionRequests) {
        actionRequests
                .to(ctx.actionTopicNamer.apply(TopicTypes.ActionTopic.request),
                        Produced.with(ctx.aSerdes.uuid(), ctx.aSerdes.request()));
    }

    public static <A> void sagaState(SagaContext<A> ctx, KStream<UUID, Saga<A>> sagaState) {
        sagaState.to(ctx.sagaTopicNamer.apply(TopicTypes.SagaTopic.state),
                Produced.with(ctx.sSerdes.uuid(), ctx.sSerdes.state()));
    }

    public static <A> void sagaStateTransitions(SagaContext<A> ctx, KStream<UUID, SagaStateTransition<A>>... transitions) {
        for (KStream<UUID, SagaStateTransition<A>> t : transitions) {
            t.to(ctx.sagaTopicNamer.apply(TopicTypes.SagaTopic.stateTransition),
                    Produced.with(ctx.sSerdes.uuid(), ctx.sSerdes.transition()));
        }
    }

    public static <A> void sagaResponses(SagaContext<A> ctx, KStream<UUID, SagaResponse> sagaResponse) {
        sagaResponse.to(ctx.sagaTopicNamer.apply(TopicTypes.SagaTopic.response),
                Produced.with(ctx.sSerdes.uuid(), ctx.sSerdes.response()));
    }
}
